AzureMonitor: Migrate Metrics query editor to React

joshhunt opened the pull request:
**What this PR does / why we need it**:

Initial React UI for the query editor, for just the Metrics query type.

![image](https://user-images.githubusercontent.com/46142/110622505-764f0d80-8193-11eb-9f3a-3b0485287403.png)


 - [x] Connect all fields
 - [x] Clear fields when their dependencies change (e.g. when namespace changes, selected metric should be cleared out) 
 - [x] Actually run the query on change
 - [x] Lay out fields properly
 - [x] Hide Angular interface
 - [x] Test dimension filters - I havent been able to get these to populate. Not sure if it's a bug in both the old and new code, or if my data set just doesn't have them.
 
**Which issue(s) this PR fixes**: 

Fixes https://github.com/grafana/grafana/issues/31108

Note, it does not fix the other issues mentioned in the Epic https://github.com/grafana/grafana/issues/30794. They'll have their own PRs :)

joshhunt committed:
AzureMonitor: Remove anys from datasource to get the inferred type

joshhunt committed:
AzureMonitor: Cast some datasource types

TODO: we want proper types for these

joshhunt committed:
AzureMonitor: Initial react Metrics editor components

sunker commented on public/app/core/angular_wrappers.ts:
```diff
@@ -196,4 +197,11 @@ export function registerAngularDirectives() {
+  react2AngularDirective('azureMonitorQueryEditor', QueryEditor, [
+    'query',
+    ['datasource', { watchDepth: 'reference' }],
+    'replaceTemplateVariable',
```
replaceTemplateVariable is not part of the "real" react [QueryEditorProps](https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/types/datasource.ts#L343-L362) api, so I'd avoid using that. The datasource has access to the template variables, so I suggest you expose a function from the datasource that can expand template variables. This function could then be used from within the query editor components.

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor.tsx:
```diff
@@ -0,0 +1,128 @@
+    { value }: SelectableValue<AzureMonitorQuery['azureMonitor'][Key]>
+  ) {
+    // TODO: investigate empty values. Do they still need to be set?
+    // TODO: when fields change, we actually need to unset all the "lower" fields that depend on this as well
```
probably :(

sunker commented on public/app/core/angular_wrappers.ts:
```diff
@@ -196,4 +197,11 @@ export function registerAngularDirectives() {
+    'query',
+    ['datasource', { watchDepth: 'reference' }],
+    'replaceTemplateVariable',
+    'onQueryChange',
```
onChange

sunker commented on public/app/core/angular_wrappers.ts:
```diff
@@ -196,4 +197,11 @@ export function registerAngularDirectives() {
+    ['datasource', { watchDepth: 'reference' }],
+    'replaceTemplateVariable',
+    'onQueryChange',
+  ]);
```
...and add onRunQuery. You should be able to run onChange without triggering a panel refresh

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor.tsx:
```diff
@@ -0,0 +1,128 @@
+}) => {
+  const metricsMetadata = useMetricsMetadata(datasource, query, subscriptionId, replaceTemplateVariable, onQueryChange);
+
+  // Single dynamic onChange function might be a bit unwieldly. Let's see how it goes.
```
we do the same for Cloud monitoring. Not with this awesome typing though :)

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNameField.tsx:
```diff
@@ -0,0 +1,61 @@
+
+    datasource
+      .getMetricNames(
+        replaceTemplateVariable(subscriptionId),
```
could we move the expanding of the template variable to the datasource?

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNameField.tsx:
```diff
@@ -0,0 +1,61 @@
+  onChange,
+}) => {
+  const [options, setOptions] = useState<Options>([]);
+  const azureMonitorIsConfigured = datasource.azureMonitorDatasource.isConfigured();
```
I know this check wasn't your idea, but it seems to be repeated a lot and I wonder what the point is of checking if the subscription id is set. Isn't it better to just let the query to azure fail if the sub id is not provided?

joshhunt replied:
Yeah this is fair, ill modify this to do essentially the same thing, but with less duplication.

sunker reviewed (commented):
Don't have credentials setup for azure at the moment, so couldn't test the look and feel, but the code looks like a great start! Just a few comments

joshhunt committed:
start dimension fields

joshhunt committed:
replace replaceTemplateVariable with datasource.replace, and rename onQueryChange to onChange

joshhunt committed:
actually just do template variable replacement in the datasource

joshhunt committed:
don't use azureMonitorIsConfigured

joshhunt committed:
Refactors, mainly around the metric metadata

 - Convert all the metric metadata options for the Select before its set into state
 - Stop using SelectableValue because it's basically any when all the properties are optional
 - the onChange function passed to the fields now just accepts the direct value, rather than wrapped in a SelectableValue

joshhunt committed:
added proper fields, and adding and removing for DimensionFields

joshhunt committed:
Update query with Dimension changes

joshhunt committed:
Width

joshhunt committed:
subscription and query type fields

joshhunt committed:
Should be feature complete now, more or less

joshhunt commented:
whatever, lets just call this ready and review!

Doesn't really contain error handling, that's for a seperate PR (this is gettiung long in the tooth as it is!).

If there's some easy feedback on the UX I'm definitely keen to incorporate that in this PR.

joshhunt commented:
If @grafana/grafana-frontend-platform has some time, I would appreciate a review from them on this one as well.

I appreciate this PR is a bit large and probably without context, so hit me up if you have any questions :)

joshhunt committed:
fix missing import

hugohaggmark reviewed (commented):
Probably not the scope for this PR, but we've introduced new support to show standard/custom/ or the same editor as for queries for template variables.

Would be great to have an example for a data source the uses the same Query editor in both Panels/Explore/Variables and we're going to migrate all internal data source plugins to this new API some time soon. 

There are a couple of examples of API usage: TestData, Prometheus and Cloud Monitoring that you can peek at. I'm glad to help answer any questions.

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx:
```diff
@@ -0,0 +1,84 @@
+      // TODO: how much of this is needed?
+      // let newSubscription: string;
+
+      // if (!query.subscription && query.queryType === 'Azure Monitor') {
```
Good question..I don't know!

joshhunt replied:
at least thanks for the reminder to look into this more!

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/index.tsx:
```diff
@@ -0,0 +1,99 @@
+import { AzureMonitorQuery } from '../../types';
+import { useMetricsMetadata } from '../metrics';
+import SubscriptionField from '../SubscriptionField';
+import MetricNamespaceField from './MetricNamespaceField';
```
Oh so this was already the index file. Never mind!

sunker reviewed (commented):
Looks great! I'd like to do some manual testing, but still don't have any credentials. Can you help me with that? 

Even though the ux does not have to be perfect at this point, I think we should be able to make some minor improvements. Let's discuss this in the daily. 

Also, are you planning to add unit tests?

joshhunt committed:
fix lint issues

joshhunt committed:
set default subscription ID

joshhunt committed:
Starting to write some tests

joshhunt committed:
tests for query editor

joshhunt committed:
Remove subscription ID from the label in Metrics

But we keep it there for the angular stuff

joshhunt committed:
MetricsQueryEditor tests

joshhunt committed:
Update index.test.tsx

joshhunt committed:
fix tests

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/AggregationField.tsx:
```diff
@@ -0,0 +1,43 @@
+}
+
+const AggregationField: React.FC<AggregationFieldProps> = ({ query, onQueryChange, aggregationOptions }) => {
+  const handleChange = useCallback(
```
not sure if I'm missing something here, but what's the real benefit of wrapping this function in useCallback?

joshhunt replied:
useCallback will memoize the function, so it won't 'change' on every render, potentially ca causing unnessisary renders.

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts:
```diff
@@ -50,6 +50,8 @@ export default class Datasource extends DataSourceApi<AzureMonitorQuery, AzureDa
   }
 
   query(options: DataQueryRequest<AzureMonitorQuery>): Observable<DataQueryResponse> {
+    console.log('datasource query');
```
remove this plz

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/types.ts:
```diff
@@ -76,6 +77,7 @@ export interface ApplicationInsightsQuery {
   metricName: string;
   timeGrainUnit: string;
   timeGrain: string;
+  // timeGrainCount: string;
```
remove or uncomment

sunker reviewed (changes requested):
Great work with the tests! 

One big thing that we've forgotten is the template variables. At the moment they're not displayed as options in the drop downs, and it's not possible to add custom values. You can take a look at how it's done [here](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/cloudwatch/components/Stats.tsx#L42). [Here](https://grafana.com/docs/grafana/latest/datasources/azuremonitor/#create-template-variables-for-metrics) are instructions on how to setup template variables for your dashboard. 

Also for the query field, error handling need to be either implemented or the catch statement should be removed.

joshhunt commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/response_parser.ts:
```diff
@@ -57,13 +57,17 @@ export default class ResponseParser {
       };
```
yeah, i kind of want to do a rewrite of many of the parts to be honest.

joshhunt committed:
add template variables to dropdowns

joshhunt commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/Field.tsx:
```diff
@@ -0,0 +1,44 @@
+    <InlineField
+      className={css`
+         {
+          align-items: center;
```
no you're right. it was some copy-paste that I just didn't think about.

joshhunt committed:
clean up

joshhunt committed:
update tests

joshhunt committed:
Reorganise react components

joshhunt committed:
Group query fields into rows

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/metrics.ts:
```diff
@@ -0,0 +1,85 @@
+          },
+        });
+
+        // TODO: Move the aggregationTypes and timeGrain defaults into `getMetricMetadata`
```
is this something that can be done now?

joshhunt replied:
I did it!

sunker reviewed (commented):
Looks great! Tested and it works fine. Also tested template variables. 

Just one questions. All pickers have empty catch statements for error handling. What do we want to do there? I don't know how/if this was handled before.

joshhunt committed:
Rename Option type, add Azure response type

joshhunt committed:
Refactor Metrics metric metadata

 - Types the Azure API
 - Moves default metadata values into datasource

joshhunt committed:
nit

joshhunt committed:
update test

sunker reviewed (approved):
LGTM
